kbfsgit/runner_test.go

```diff
@@ -666,3 +666,210 @@ func TestPushcertOptions(t *testing.T) {
+	dotgit1 := filepath.Join(git1, ".git")
+
+	// Make shared repo with 2 branches.  Make sure there's an initial
+	// pack-refs file.
+	makeLocalRepoWithOneFile(t, git1, "foo", "hello", "")
+	gitExec(t, dotgit1, git1, "pack-refs", "--all")
+	testPushWithTemplate(t, ctx, config, git1,
+		[]string{"refs/heads/master:refs/heads/master"},
+		"ok %s\n\n", "user1,user2")
+	testPushWithTemplate(t, ctx, config, git1,
+		[]string{"refs/heads/master:refs/heads/test"},
+		"ok %s\n\n", "user1,user2")
+
+	// Config for the second user.
+	config2 := libkbfs.ConfigAsUser(config, "user2")
+	defer libkbfs.CheckConfigAndShutdown(ctx, t, config2)
+	tempdir2, err := ioutil.TempDir(os.TempDir(), "journal_server")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempdir2)
+	err = config2.EnableDiskLimiter(tempdir2)
+	require.NoError(t, err)
+	err = config2.EnableJournaling(
+		ctx, tempdir2, libkbfs.TLFJournalSingleOpBackgroundWorkEnabled)
+	require.NoError(t, err)
+
+	git2, err := ioutil.TempDir(os.TempDir(), "kbfsgittest")
+	require.NoError(t, err)
+	defer os.RemoveAll(git2)
+	dotgit2 := filepath.Join(git2, ".git")
+
+	heads := testListAndGetHeadsWithName(t, ctx, config2, git2,
+		[]string{"refs/heads/master", "refs/heads/test", "HEAD"}, "user1,user2")
+	require.Equal(t, heads[0], heads[1])
+
+	// Have the second user refpack.
+	r, err := newRunner(ctx, config2, "origin",
+		fmt.Sprintf("keybase://private/user1,user2/test"),
+		dotgit2, nil, nil, testErrput{t})
+	require.NoError(t, err)
+
+	// Stall it after it takes the lock.
+	packOnStalled, packUnstall, packCtx := libkbfs.StallMDOp(
+		ctx, config2, libkbfs.StallableMDAfterGetRange, 1)
+	packErrCh := make(chan error)
+	go func() {
+		packErrCh <- r.packRefs(packCtx)
+	}()
+	select {
+	case <-packOnStalled:
+	case <-ctx.Done():
+		t.Fatal(ctx.Err())
+	}
+
+	// While the second user is stalled, have the first user delete
+	// one of the refs.  Wait until it tries to get the lock, and then
+	// release the pack-refs call.
+	deleteOnStalled, deleteUnstall, deleteCtx := libkbfs.StallMDOp(
+		ctx, config, libkbfs.StallableMDGetRange, 1)
+	inputReader, inputWriter := io.Pipe()
+	defer inputWriter.Close()
+	go func() {
+		inputWriter.Write([]byte("push :refs/heads/test\n"))
+		inputWriter.Write([]byte("\n\n"))
+	}()
+
+	var output bytes.Buffer
+	deleteRunner, err := newRunner(ctx, config, "origin",
+		"keybase://private/user1,user2/test",
+		dotgit1, inputReader, &output, testErrput{t})
+	require.NoError(t, err)
+	deleteErrCh := make(chan error)
+	go func() {
+		deleteErrCh <- deleteRunner.processCommands(deleteCtx)
+	}()
+	select {
+	case <-deleteOnStalled:
+	case <-ctx.Done():
+		t.Fatal(ctx.Err())
+	}
+	// Release it, and it should block on getting the lock.
+	close(deleteUnstall)
+
+	// Now let the pack-refs finish.
+	close(packUnstall)
+	select {
+	case err := <-packErrCh:
+		require.NoError(t, err)
+	case <-ctx.Done():
+		t.Fatal(ctx.Err())
+	}
+
+	// And the delete should finish right after.
```

Comment:
This is what I was talking about. There's no mechanism that guarantees that this happens right *after* . If locking fails and it happened before `<-packErrCh` happened, this test would still pass.

Reply:
Hmm, if locking fails then `deleteErrCh` should receive a non-nil error and the test would fail.

But there is no guarantee that the delete command will actually block on the lock -- it might get delayed, and only attempt to get the lock after the pack refs is done.  Which would be a less effective test, but yeah it's too much work to tell when a lock attempt is blocking, not really worth it to get that guarantee in my opinion.

Reply:
By "locking fails", I meant we accidentally used wrong `LockID` and made it ineffective, not that the locking operations returned error.

But yeah it seems to be a lot more work to know that for sure here. Agreed!